architecture/dns.md

```diff
@@ -0,0 +1,201 @@
+## Context
+
+Sangam requires a decentralized DNS secured via a blockchain to function. This 
+feature can be implemented as a module so other applications can make use of it
+and support its maintenance. The DNS will be fairly simple, mapping some key to
+some encoded value. The DNS will support payments for registration and updates
+as well as security checks to prevent malicious behavior.  
+
+## Decision
+
+We will implement a DNS module supporting the following functionality:
+
+- basic KVStore functionality for any type of DNS entry
+- required payment amounts for setting and updating entries
+- ownership verification for entry updates
+- commit-reveal scheme to prevent validators from stealing registration rights
+- optional domain ownership over a certain key prefix
+- [future] support governance of DNS entries
+
+## Types
+
+The DNS system will consist of two types of entries: `DomainEntry` and `ContentEntry`.
+
+A domain may contain any number of subdomains and content entries. Both domains and entries must
+be uniquely namespaced under its subdomain (or globally unique if it is a top-level domain).
+
+A domain is owned by a `DomainOwner` interface, which defines authentication methods 
+for adding/updating/removing content and subdomains. Each domain contains a list of subdomains, 
+a list of content entries, and a parent domain (top level domains have nil as a domain).
+
+A content entry has a single domain, and an array of content hashes, with each update to the content hash
+appended to the array. Thus one can retreive the latest version as well as any previous version from the same entry.
+
+Domains are seperated by the `.` seperator, while content is seperated by the `/` seperator. Thus, a content `blog-post-1` 
+stored under the domain hierarchy `domain1 -> domain2 -> domain3` will be stored under the full path: 
+`domain1.domain2.domain3/blog-post-1`.
+
+All domain entries are set in store under the key `{DomainPrefix}/{Domain.Path()}`. All Content entries are stored under the 
+key `{ContentPrefix}/Content.Path()`.
+
+Each content hash must also have a reverse mapping from `{ReverseContentPrefix}/{ContentHash} -> {Content.Path()}`.
+
+
+```go
+// exported.go
+type DomainEntry interface {
+    Owner() DomainOwner
+    SubDomains() []DomainEntry
+    Contents() []ContentEntry
+    // returns parent domain, nil if top-level domain
+    Parent() DomainEntry
+
+    // returns the string identifier of this domain
+    String() string
+    // returns the full path string representation
+    // up to the global top-level domain
+    Path() string
+
+    // Returns the subdomain under this domain as specified by the given string
+    GetSubDomain(path string) DomainEntry
```

Comment:
what is the `path`? the full path, the relative path including the current domain or the path starting at the sub-domain that should exist at this domain?